src/api/system.c

```diff
@@ -638,6 +638,110 @@ static int f_set_window_opacity(lua_State *L) {
 }
 
 
+
+typedef struct {
+  const char* symbol;
+  void* address;
+} lua_function_node;
+
+#define P(FUNC) { "lua_" #FUNC, (void*)(lua_##FUNC) }
+static void* api_require(const char* symbol) {
+  static lua_function_node nodes[] = {
+    P(absindex), P(arith), P(atpanic), P(callk), P(checkstack), 
+    P(close), P(compare), P(concat), P(copy), P(createtable), P(dump), 
+    P(error),  P(gc), P(getallocf), P(getctx), P(getfield), P(getglobal),
+    P(gethook), P(gethookcount), P(gethookmask), P(getinfo), P(getlocal),
+    P(getmetatable), P(getstack), P(gettable), P(gettop), P(getupvalue),
+    P(getuservalue), P(insert), P(isnumber), 
+    P(isstring), P(isuserdata), P(len), P(load), 
+    P(newstate), P(newthread), P(newuserdata), P(next), 
+    P(pcallk), P(pushboolean), P(pushcclosure), 
+    P(pushfstring), P(pushinteger), P(pushlightuserdata),
+    P(pushlstring), P(pushnil), P(pushnumber), P(pushstring),
+    P(pushthread), P(pushunsigned), P(pushvalue), P(pushvfstring), P(rawequal),
+    P(rawget), P(rawgeti), P(rawgetp), P(rawlen), P(rawset), P(rawseti),
+    P(rawsetp), P(remove), P(replace), P(resume), P(setallocf), 
+    P(setfield), P(setglobal), P(sethook), P(setlocal), P(setmetatable), 
+    P(settable), P(settop), P(setupvalue), P(setuservalue), P(status), 
+    P(tocfunction), P(tointegerx), P(tolstring), P(toboolean),
+    P(tonumberx), P(topointer), P(tothread), 
+    P(tounsignedx), P(touserdata), P(type), P(typename), P(upvalueid),
+    P(upvaluejoin), P(version), P(xmove), P(yieldk)
+  };
+  for (int i = 0; i < sizeof(nodes) / sizeof(lua_function_node); ++i) {
+    if (strcmp(nodes[i].symbol, symbol) == 0)
+      return nodes[i].address;
+  }
+  return NULL;
+}
+
+static int loader_plugin(lua_State* L) {
+  size_t sname, modlen, pathlen;
+  char buffer[512] = "lua_open_";
+  const char* modname = luaL_checklstring(L, -2, &modlen);
+  const char* path = luaL_checklstring(L, -1, &pathlen);
+  void* library = SDL_LoadObject(path);
+  if (modlen == 0 || !library)
+    return luaL_error(L, "Unable to load %s: %s", modname, SDL_GetError());
+  for (sname = modlen - 1; sname > 0 && modname[sname] != '.'; --sname);
+  strcat(buffer, &modname[sname+1]);
+  int (*entrypoint)(lua_State* L, void*) = SDL_LoadFunction(library, buffer);
+  if (!entrypoint) {
+    return luaL_error(L, "Unable to load %s: Can't find entrypoint. Requires a "
+      "function defined as int %s(lua_State* L, void* (*symbol)(const char*))", 
+    modname, buffer);
+  }
+  lua_pushlightuserdata(L, library);
+  lua_setfield(L, LUA_REGISTRYINDEX, modname);
+  if (entrypoint(L, api_require) == 0)
+    return luaL_error(L, "Unable to load %s: Your entrypoint must return at\
+      least one argument.");
+  return 1;
+}
+
+#if _WIN32
+  #define PATH_SEPARATOR '\\'
+  #define DYNAMIC_SUFFIX "dll"
+#else 
+  #if __APPLE__
+    #define DYNAMIC_SUFFIX "lib"
+  #else
+    #define DYNAMIC_SUFFIX "so"
+  #endif
+  #define PATH_SEPARATOR '/'
+#endif
```

Comment:
The preprocessors _WIN32 and \_\_APPLE\_\_ should usually be defined but in the case they are value that has a false truthiness we might fall through to the Unix Logic.

 ```suggestion
#if defined(_WIN32)
  #define PATH_SEPARATOR '\\'
  #define DYNAMIC_SUFFIX "dll"
#elif defined(__APPLE__)
  #define PATH_SEPARATOR '/'
  #define DYNAMIC_SUFFIX "lib"
#elif defined(__linux__)
  #define PATH_SEPARATOR '/'
  #define DYNAMIC_SUFFIX "so"
#endif
#endif
```

Reply:
Right, we may set an preprocessor #error directive in the unknown OS branch. Otherwise we may default to unix or check for other unix-like Os like BSD systems.

Reply:
I'm happy to make both these changes, but this does not follow the conventions elsewhere in lite (not bothering with defined, not bothering to explicitly check linux). If we change it here, we should change it everywhere else that's applicable.

Personally, I don't think it's necessary to have `defined`; though I don't mind adding it (is there ever a realistic case where these would have different outcomes)?

Regarding defaulting to linux, I think that's a reasonable step, as there are far more *nix flavours than just linux, and they're all relatively similar in terms of this kind of thing.